File: Java Enrolment System - Basic FIT5136/src/StudentEnrollmentSystem.java
import java.util.*;

/**
 * Student Enrollment System - Manages students and units within the Student Enrollment System and provides a menu-driven interface.
 * <p>
 * This system allows administrators to manage students and units, and enables students to enroll or remove themselves
 * from units. It includes features for adding/removing students and units, listing all students and units, and handling
 * enrollment with confirmation prompts.
 * </p>
 *
 * <p>
 * @author Antoni Erdeg - antoni.erdeg@monash.edu<br>
 * @version: 1.0
 * </p>
 */
public class StudentEnrollmentSystem {
    // All final variables are constants and should be upper case.
    // Refer to PartTimeStudent for non-constant variable example i.e. courseAdvisor
    private final Map<String, Student> STUDENTS = new HashMap<>();
    private final Map<String, Unit> UNITS = new HashMap<>();
    private final Scanner SCANNER = new Scanner(System.in);
    private Administrator admin;

    /**
     * Initializes the Student Enrollment System with an administrator.
     *
     * @param admin The administrator managing the system.
     */
    public StudentEnrollmentSystem(Administrator admin) {
        this.admin = admin;
    }

    /**
     * Displays a standard header before each menu.
     *
     * @param pageName The name of the current page to display in the header.
     */
    public void displayHeader(String pageName) {
        String header = "\n---------------------------------\n" +
                "|         FIT5136 Demo           |\n" +
                "|    Student Enrolment System    |\n" +
                "---------------------------------\n" +
                " -- Current Page: " + pageName + " --";

        System.out.println(header);
    }

    /**
     * Retrieves the map of Students in the system.
     *
     * @return A map of student IDs to Student objects.
     */
    public Map<String, Student> getStudents() {
        return STUDENTS;
    }

    /**
     * Adds a student to the system if the student ID does not already exist.
     *
     * @param student The student to be added.
     */
    public void addStudent(Student student) {
        String studentId = student.getId();
        if (STUDENTS.containsKey(studentId)) {
            System.out.println("Error: Student ID '" + studentId + "' already exists. Cannot add new student with this ID.");
        } else {
            STUDENTS.put(studentId, student);
            System.out.println("Student '" + studentId + "' added successfully.");
        }
    }


    /**
     * Removes a student from the system by their ID if they exist.
     *
     * @param studentId The ID of the student to be removed.
     */
    public void removeStudent(String studentId) {
        if (studentExists(studentId)) {
            STUDENTS.remove(studentId);
            System.out.println("Student with ID '" + studentId + "' has been removed.");
        } else {
            System.out.println("Unable to remove student with ID '" + studentId + "' as they do not exist.");
        }
    }

    /**
     * Checks if a student exists in the system by their ID.
     *
     * @param studentId The ID of the student.
     * @return True if the student exists, false otherwise.
     */
    public boolean studentExists(String studentId) {
        return STUDENTS.containsKey(studentId);
    }

    /**
     * Adds a unit to the system if the unit name does not already exist.
     *
     * @param unit The unit to be added.
     */
    public void addUnit(Unit unit) {
        String unitName = unit.getCode();
        if (UNITS.containsKey(unitName)) {
            System.out.println("Error: Unit name '" + unitName + "' already exists. Cannot add new unit with this name.");
        } else {
            UNITS.put(unitName, unit);
            System.out.println("Unit '" + unitName + "' added successfully.");
        }
    }


    /**
     * Checks if a unit is available in the system by its name.
     *
     * @param unit The unit to check.
     * @return True if the unit is available, false otherwise.
     */
    public boolean isUnitAvailable(Unit unit) {
        return UNITS.containsKey(unit.getCode());
    }

    /**
     * Adds a unit to a student's list of units with confirmation.
     *
     * @param studentId The ID of the student.
     * @param unit The unit to be added.
     */
    public void addUnitToStudent(String studentId, Unit unit) {
        if (studentExists(studentId)) { // Check if student exists
            Student student = STUDENTS.get(studentId);

            // Check if student is already enrolled in the unit
            if (student.getUnits().contains(unit)) {
                System.out.println("Student is already enrolled in the unit '" + unit.getCode() + "'.");
                return;
            }

            // Check if student is already enrolled in 4 units
            if (student.getUnits().size() >= 4) {
                System.out.println("Student is already enrolled in the maximum number of units (4 units).");
                return;
            }

            System.out.print("Confirm enrollment in unit '" + unit.getCode() + "' (Y/N): ");
            String confirmation = SCANNER.nextLine().trim().toUpperCase();

            if (confirmation.equals("Y")) { // Confirm action
                student.addUnit(unit, this);
                System.out.println("\nEnrolled in unit '" + unit.getCode() + "' successfully.");
            } else {
                System.out.println("\nEnrollment in unit '" + unit.getCode() + "' cancelled.");
            }
        } else {
            System.out.println("\nStudent not found.");
        }
    }

    /**
     * Removes a unit from a student's list of units with confirmation.
     *
     * @param studentId The ID of the student.
     * @param unitName The name of the unit to be removed.
     */
    public void removeUnitFromStudent(String studentId, String unitName) {
        if (studentExists(studentId)) { // Check if student exists
            Student student = STUDENTS.get(studentId);
            Set<Unit> enrolledUnits = student.getUnits();

            // Ensure the set contains the unit by its name
            Unit unitToRemove = null;
            for (Unit unit : enrolledUnits) {
                if (unit.getCode().equals(unitName)) {
                    unitToRemove = unit;
                    break;
                }
            }

            if (unitToRemove != null) { // Unit found
                System.out.print("Confirm removal of unit '" + unitName + "' (Y/N): ");
                String confirmation = SCANNER.nextLine().trim().toUpperCase();

                if (confirmation.equals("Y")) { // Confirm action
                    student.removeUnit(unitName);
                } else {
                    System.out.println("Removal of unit '" + unitName + "' cancelled.");
                }
            } else {
                System.out.println("Student isn't currently enrolled in the unit '" + unitName + "'.");
            }
        } else {
            System.out.println("Student not found.");
        }
    }

    /**
     * Lists all units currently enrolled by a student.
     *
     * @param studentId The ID of the student.
     */
    public void listStudentUnits(String studentId) {
        if (studentExists(studentId)) { // Check if student exists
            Student student = STUDENTS.get(studentId);
            student.listUnits();
        } else {
            System.out.println("Student not found.");
        }
    }

    /**
     * Retrieves all units in the system.
     *
     * @return A collection of units.
     */
    public Collection<Unit> getUnits() {
        return UNITS.values();
    }

    /**
     * Displays the main menu and handles user interactions.
     */
    public void displayMainMenu() {
        while (true) {
            displayHeader("Main Menu");
            System.out.println("\nSelect User Type:");
            System.out.println("1. Student");
            System.out.println("2. Administrator");
            System.out.println("0. Exit");
            System.out.print("Enter choice: ");
            int choice = SCANNER.nextInt();
            SCANNER.nextLine();

            switch (choice) {
                case 1:
                    studentMenu(); // Show student menu
                    break;
                case 2:
                    adminMenu(); // Show administrator menu
                    break;
                case 0:
                    System.out.println("Exiting system.");
                    System.exit(0);
                default:
                    System.out.println("Invalid choice."); // Handle invalid choice
            }
        }
    }

    /**
     * Provides the menu for student interactions.
     */
    private void studentMenu() {
        System.out.print("Enter Student ID: ");
        String studentId = SCANNER.nextLine();

        if (!studentExists(studentId)) { // Check if student exists
            System.out.println("Student does not exist in the enrollment system.");
            return;
        }

        while (true) {
            displayHeader("Student Menu");
            System.out.println("\nStudent Menu:");
            System.out.println("1. Add Unit");
            System.out.println("2. Remove Unit");
            System.out.println("3. List Current Units");
            System.out.println("4. Back to Main Menu");
            System.out.println("0. Exit");
            System.out.print("Enter choice: ");
            int studentChoice = SCANNER.nextInt();
            SCANNER.nextLine();

            switch (studentChoice) {
                case 1:
                    // Show available units and handle unit enrollment
                    this.displayHeader("Student Menu: Add Unit");
                    System.out.println("\nAvailable Units:"+"\n-----------------");
                    getUnits().forEach(unit -> System.out.println(unit.toString()));
                    System.out.print("\nEnter Unit Code to Enroll: ");
                    String unitName = SCANNER.nextLine();
                    Unit unitToAdd = UNITS.get(unitName);
                    if (unitToAdd != null) { // Check if unit exists
                        addUnitToStudent(studentId, unitToAdd);
                    } else {
                        System.out.println("Unit '" + unitName + "' not found.");
                    }
                    break;
                case 2:
                    // Show enrolled units and handle unit removal
                    this.displayHeader("Student Menu: Remove Unit");
                    System.out.println("\nUnits Enrolled:"+"\n-----------------");
                    listStudentUnits(studentId);

                    Student student = STUDENTS.get(studentId);
                    if (student == null || student.getUnits().isEmpty()) { // Check if student has units
                        System.out.println("Not enrolled in any Units.");
                        break; // Return to student menu
                    }

                    System.out.print("\nEnter Unit Code to Remove: ");
                    String unitNameToRemove = SCANNER.nextLine();

                    removeUnitFromStudent(studentId, unitNameToRemove);

                    break;
                case 3:
                    // List current units for the student
                    System.out.println("\nCurrent Units:"+"\n-----------------");
                    listStudentUnits(studentId);

                    // Prompt to return to the menu
                    System.out.println("\nPress any key or 'B' to return to the Menu.");
                    SCANNER.nextLine(); // Wait for user input
                    break;
                case 4:
                    return; // Return to main menu
                case 0:
                    System.out.println("Exiting system.");
                    System.exit(0);
                default:
                    System.out.println("Invalid choice."); // Handle invalid choice
            }
        }
    }

    /**
     * Provides the menu for administrator interactions.
     */
    private void adminMenu() {
        while (true) {
            displayHeader("Administrator Menu");
            System.out.println("\nAdministrator Menu:");
            System.out.println("1. Add Student");
            System.out.println("2. Remove Student");
            System.out.println("3. Add Unit");
            System.out.println("4. Remove Unit");
            System.out.println("5. List All Students");
            System.out.println("6. List All Units");
            System.out.println("7. Back to Main Menu");
            System.out.println("8. Add Unit to Student");  // New option
            System.out.println("9. Remove Unit from Student");  // New option
            System.out.println("0. Exit");
            System.out.print("Enter choice: ");
            int adminChoice = SCANNER.nextInt();
            SCANNER.nextLine();  // Consume newline

            switch (adminChoice) {
                case 1:
                    // Add a new student
                    displayHeader("Admin Menu: Add New Student");
                    System.out.print("Enter New Student Name: ");
                    String newStudentName = SCANNER.nextLine();
                    System.out.print("Enter New Student ID: ");
                    String newStudentId = SCANNER.nextLine();
                    System.out.print("Enter New Student Address: ");
                    String newStudentAddress = SCANNER.nextLine();
                    System.out.print("Enter New Student Phone: ");
                    String newStudentPhone = SCANNER.nextLine();
                    System.out.print("Enter New Student Email: ");
                    String newStudentEmail = SCANNER.nextLine();
                    System.out.print("Enter Type of Student - [F] Full-time or [P] Part-time: ");
                    String newStudentType = SCANNER.nextLine().trim().toUpperCase();

                    Student newStudent;
                    if (newStudentType.equals("F")) {
                        newStudent = new FullTimeStudent(newStudentId, newStudentName, newStudentAddress, newStudentPhone, newStudentEmail);
                    } else if (newStudentType.equals("P")) {
                        newStudent = new PartTimeStudent(newStudentId, newStudentName, newStudentAddress, newStudentPhone, newStudentEmail);
                    } else {
                        System.out.println("Invalid Option. Returning to Main Menu.");
                        break;
                    }
                    addStudent(newStudent);
                    break;

                case 2:
                    // Remove a student
                    displayHeader("Admin Menu: Remove Student");
                    System.out.println("\nCurrent Students:" + "\n-----------------");
                    List<Student> studentList = new ArrayList<>(STUDENTS.values());
                    if (studentList.isEmpty()) {
                        System.out.println("No students currently enrolled.");
                    } else {
                        for (Student student : studentList) {
                            System.out.println(student);
                        }
                        System.out.print("\nEnter Student ID to Remove: ");
                        String studentIdToRemove = SCANNER.nextLine();
                        System.out.print("Confirm removal of student '" + studentIdToRemove + "' (Y/N): ");
                        String confirmation = SCANNER.nextLine().trim().toUpperCase();
                        if (confirmation.equals("Y")) {
                            removeStudent(studentIdToRemove);
                        } else {
                            System.out.println("Operation cancelled.");
                        }
                    }
                    break;

                case 3:
                    // Add a new unit
                    displayHeader("Admin Menu: Add New Unit");
                    System.out.print("Enter New Unit Code: ");
                    String newUnitName = SCANNER.nextLine();
                    System.out.print("Enter New Unit Title: ");
                    String newUnitTitle = SCANNER.nextLine();
                    System.out.print("Enter Credit Points: ");
                    int newUnitPoints;
                    try {
                        newUnitPoints = Integer.parseInt(SCANNER.nextLine());
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid input for credit points. Returning to Main Menu.");
                        break;
                    }

                    Unit newUnit = new Unit(newUnitName, newUnitTitle, newUnitPoints);
                    System.out.print("Confirm creation of unit '" + newUnitName + "' to the system (Y/N): ");
                    String addUnitConfirmation = SCANNER.nextLine().trim().toUpperCase();
                    if (addUnitConfirmation.equals("Y")) {
                        addUnit(newUnit);
                    } else {
                        System.out.println("Operation cancelled.");
                    }
                    break;

                case 4:
                    // Remove a unit
                    displayHeader("Admin Menu: Remove Unit");
                    System.out.println("\nCurrent Units:" + "\n-----------------");
                    List<Unit> unitList = new ArrayList<>(UNITS.values());
                    if (unitList.isEmpty()) {
                        System.out.println("No units currently available.");
                    } else {
                        for (Unit unit : unitList) {
                            System.out.println(unit);
                        }
                    }
                    System.out.print("\nEnter Unit Code to Remove: ");
                    String unitNameToRemove = SCANNER.nextLine().trim();
                    if (UNITS.containsKey(unitNameToRemove)) {
                        System.out.print("Confirm removal of unit '" + unitNameToRemove + "' (Y/N): ");
                        String removeUnitConfirmation = SCANNER.nextLine().trim().toUpperCase();
                        if (removeUnitConfirmation.equals("Y")) {
                            UNITS.remove(unitNameToRemove);
                            System.out.println("Unit '" + unitNameToRemove + "' removed successfully.");
                        } else {
                            System.out.println("Operation cancelled.");
                        }
                    } else {
                        System.out.println("Error: Unit '" + unitNameToRemove + "' does not exist. Cannot remove a non-existent unit.");
                    }
                    break;

                case 5:
                    // List all students
                    displayHeader("Admin Menu: List All Students");
                    System.out.println("\nCurrent Students:" + "\n-----------------");
                    studentList = new ArrayList<>(STUDENTS.values());
                    if (studentList.isEmpty()) {
                        System.out.println("No students currently enrolled.");
                    } else {
                        for (Student student : studentList) {
                            System.out.println(student);
                        }
                    }
                    System.out.println("\nPress any key or 'B' to return to the Menu.");
                    SCANNER.nextLine();  // Wait for user input
                    break;

                case 6:
                    // List all units
                    displayHeader("Admin Menu: List All Units");
                    System.out.println("\nCurrent Units:" + "\n-----------------");
                    unitList = new ArrayList<>(UNITS.values());
                    if (unitList.isEmpty()) {
                        System.out.println("No units currently available.");
                    } else {
                        for (Unit unit : unitList) {
                            System.out.println(unit);
                        }
                    }
                    System.out.println("\nPress any key or 'B' to return to the Menu.");
                    SCANNER.nextLine();  // Wait for user input
                    break;

                case 7:
                    return;  // Return to main menu

                case 8:
                    // Add Unit to Student
                    displayHeader("Admin Menu: Add Unit to Student");
                    System.out.println("\nCurrent Students:" + "\n-----------------");
                    studentList = new ArrayList<>(STUDENTS.values());
                    if (studentList.isEmpty()) {
                        System.out.println("No students currently enrolled.");
                        return;
                    } else {
                        for (Student student : studentList) {
                            System.out.println(student);
                        }
                    }
                    System.out.print("\nStudent ID: ");
                    String studentId = SCANNER.nextLine().trim();
                    if (studentExists(studentId)) {
                        System.out.println("\nAvailable Units:" + "\n-----------------");
                        for (Unit unit : UNITS.values()) {
                            System.out.println(unit);
                        }
                        System.out.print("\nEnter Unit Code to Enroll: ");
                        String unitName = SCANNER.nextLine().trim();
                        Unit unitToAdd = UNITS.get(unitName);
                        if (unitToAdd != null) {
                            addUnitToStudent(studentId, unitToAdd);
                        } else {
                            System.out.println("Unit '" + unitName + "' not found.");
                        }
                    } else {
                        System.out.println("Student not found.");
                    }
                    break;

                case 9:
                    // Remove Unit from Student
                    displayHeader("Admin Menu: Remove Unit from Student");
                    System.out.println("\nCurrent Students:" + "\n-----------------");
                    studentList = new ArrayList<>(STUDENTS.values());
                    if (studentList.isEmpty()) {
                        System.out.println("No students currently enrolled.");
                        return;
                    } else {
                        for (Student student : studentList) {
                            System.out.println(student);
                        }
                    }
                    System.out.print("\nEnter Student ID: ");
                    studentId = SCANNER.nextLine().trim();
                    if (studentExists(studentId)) {
                        System.out.println("\nStudent's Current Units:"+"\n-----------------");
                        listStudentUnits(studentId);
                        if( !STUDENTS.get(studentId).getUnits().isEmpty() ){
                            System.out.print("\nEnter Unit Code to Remove: ");
                            String unitNameToRemoveFromStudent = SCANNER.nextLine().trim();
                            removeUnitFromStudent(studentId, unitNameToRemoveFromStudent);
                        }
                        else
                        {
                            System.out.println("\nThis student is not enrolled in any units.");
                            return;
                        }

                    } else {
                        System.out.println("Student not found.");
                    }
                    break;

                case 0:
                    System.out.println("Exiting system.");
                    System.exit(0);

                default:
                    System.out.println("Invalid choice.");  // Handle invalid choice
            }
        }
    }


    /**
     * Entry point of the application.
     *
     * @param args Command-line arguments (not used).
     */
    public static void main(String[] args) {
        Administrator admin = new Administrator("John");
        StudentEnrollmentSystem system = new StudentEnrollmentSystem(admin);
        system.addStudent(new FullTimeStudent("123123", "John Lee", "123 Fake Street", "1111111111", "JohnLee@fakestreet.com"));
        system.addUnit(new Unit("FIT5136", "Software Engineering", 6));
        system.displayMainMenu();
    }
}

File: Python Enrolment System - Basic FIT5136/src/StudentEnrollmentSystem.py
from Administrator import Administrator
from FullTimeStudent import FullTimeStudent
from PartTimeStudent import PartTimeStudent
from Unit import Unit

import sys

"""
Student Enrollment System - Manages students and units within the Student Enrollment System and provides a menu-driven interface.

This system allows administrators to manage students and units, and enables students to enroll or remove themselves
from units. It includes features for adding/removing students and units, listing all students and units, and handling
enrollment with confirmation prompts.

Author: Antoni Erdeg - antoni.erdeg@monash.edu
Version: 1.0
"""

class StudentEnrollmentSystem:
    def __init__(self, admin):
        """
        Initializes the Student Enrollment System with an administrator.

        :param admin: The administrator managing the system.
        """
        self.students = {}  # Map of student IDs to Student objects
        self.units = {}  # Map of unit names to Unit objects
        self.admin = admin

    def display_header(self, page_name):
        """
        Displays a standard header before each menu.

        :param page_name: The name of the current page to display in the header.
        """
        header = (
            "\n---------------------------------\n"
            "|         FIT5136 Demo           |\n"
            "|    Student Enrollment System   |\n"
            "---------------------------------\n"
            f" -- Current Page: {page_name} --"
        )
        print(header)

    def get_students(self):
        """
        Retrieves the map of students in the system.

        :return: A dictionary of student IDs to Student objects.
        """
        return self.students

    def get_student(self, studentId):
        """
        Retrieves a Student object from the system based on the provided student ID.

        :param studentId: The ID of the student to retrieve.
        :return: The Student object corresponding to the provided student ID, or None if not found.
        """
        return self.students.get(studentId)

    def add_student(self, student):
        """
        Adds a student to the system if the student ID does not already exist.

        :param student: The student to be added.
        """
        student_id = student.get_id()
        if student_id in self.students:
            print(f"Error: Student ID '{student_id}' already exists. Cannot add new student with this ID.")
            return

        self.students[student_id] = student
        print(f"Student '{student_id}' added successfully.")


    def remove_student(self, student_id):
        """
        Removes a student from the system by their ID if they exist.

        :param student_id: The ID of the student to be removed.
        """
        if self.student_exists(student_id):
            self.students.pop(student_id, None)
            print(f"Student with ID '{student_id}' has been removed.")
        else:
            print(f"Unable to remove student with ID '{student_id}' as they do not exist.")

    def student_exists(self, student_id):
        """
        Checks if a student exists in the system by their ID.

        :param student_id: The ID of the student.
        :return: True if the student exists, False otherwise.
        """
        return student_id in self.students

    def add_unit(self, unit):
        """
        Adds a unit to the system if the unit name does not already exist.

        :param unit: The unit to be added.
        """
        unit_name = unit.get_code()
        if unit_name in self.units:
            print(f"Error: Unit name '{unit_name}' already exists. Cannot add new unit with this name.")
        else:
            self.units[unit_name] = unit
            print(f"Unit '{unit_name}' added successfully.")


    def is_unit_available(self, unit):
        """
        Checks if a unit is available in the system by its name.

        :param unit: The unit to check.
        :return: True if the unit is available, False otherwise.
        """
        return unit.get_code() in self.units

    def add_unit_to_student(self, student_id, unit):
        """
        Adds a unit to a student's list of units with confirmation.

        :param student_id: The ID of the student.
        :param unit: The unit to be added.
        """
        if self.student_exists(student_id):  # Check if student exists
            student = self.students[student_id]

            # Check if student is already enrolled in the unit
            if unit in student.get_units():
                print(f"Student is already enrolled in the unit '{unit.get_code()}'.")
                return

            # Check if student is already enrolled in 4 units
            if len(student.get_units()) >= 4:
                print("Student is already enrolled in the maximum number of units (4 units).")
                return

            confirmation = input(f"Confirm enrollment in unit '{unit.get_code()}' (Y/N): ").strip().upper()
            if confirmation == "Y":  # Confirm action
                student.add_unit(unit, self)
                print(f"\nEnrolled in unit '{unit.get_code()}' successfully.")
            else:
                print(f"\nEnrollment in unit '{unit.get_code()}' cancelled.")
        else:
            print("Student not found.")

    def remove_unit_from_student(self, student_id, unit_name):
        """
        Removes a unit from a student's list of units with confirmation.

        :param student_id: The ID of the student.
        :param unit_name: The name of the unit to be removed.
        """
        if self.student_exists(student_id):  # Check if student exists
            student = self.students[student_id]
            unit_to_remove = None

            # Ensure the set contains the unit by its name
            for unit in student.get_units():
                if unit.get_code() == unit_name:
                    unit_to_remove = unit
                    break

            if unit_to_remove is not None:  # Unit found
                confirmation = input(f"Confirm removal of unit '{unit_name}' (Y/N): ").strip().upper()
                if confirmation == "Y":  # Confirm action
                    student.remove_unit(unit_name)
                else:
                    print(f"Removal of unit '{unit_name}' cancelled.")
            else:
                print(f"Student isn't currently enrolled in the unit '{unit_name}'.")
        else:
            print("Student not found.")

    def list_student_units(self, student_id):
        """
        Lists all units currently enrolled by a student.

        :param student_id: The ID of the student.
        """
        if self.student_exists(student_id):  # Check if student exists
            student = self.students[student_id]
            student.list_units()
        else:
            print("Student not found.")

    def get_units(self):
        """
        Retrieves all units in the system.

        :return: A collection of units.
        """
        return self.units.values()

    def display_main_menu(self):
        """
        Displays the main menu and handles user interactions.
        """
        while True:
            self.display_header("Main Menu")
            print("\nSelect User Type:")
            print("1. Student")
            print("2. Administrator")
            print("0. Exit")
            choice = input("Enter choice: ").strip()

            if choice == '1':
                self.student_menu()  # Show student menu
            elif choice == '2':
                self.admin_menu()  # Show administrator menu
            elif choice == '0':
                print("Exiting system.")
                sys.exit()
            else:
                print("Invalid choice.")  # Handle invalid choice

    def student_menu(self):
        """
        Provides the menu for student interactions.
        """
        student_id = input("Enter Student ID: ").strip()

        if not self.student_exists(student_id):  # Check if student exists
            print("Student does not exist in the enrollment system.")
            return

        while True:
            self.display_header("Student Menu")
            print("\nStudent Menu:")
            print("1. Add Unit")
            print("2. Remove Unit")
            print("3. List Current Units")
            print("4. Back to Main Menu")
            print("0. Exit")
            student_choice = input("Enter choice: ").strip()

            if student_choice == '1':
                self.display_header("Student Menu: Add Unit")
                print("\nAvailable Units:" + "\n-----------------")

                # Display the list of available units
                for unit in self.get_units():
                    print(unit)

                # Prompt for the unit name to enroll
                unit_name = input("\nEnter Unit Code to Enroll: ").strip()
                unit_to_add = self.units.get(unit_name)  # Retrieve the unit object

                # Check if the unit exists
                if unit_to_add:
                    self.add_unit_to_student(student_id, unit_to_add)  # Enroll the student in the unit
                else:
                    print(f"Unit '{unit_name}' not found.")  # Print error message if unit is not found

            elif student_choice == '2':
                self.display_header("Student Menu: Remove Unit")
                print("\nUnits Enrolled:" + "\n-----------------")

                # Display the list of enrolled units for the student
                self.list_student_units(student_id)

                student = self.get_student(student_id)  # Retrieve the student object

                # Check if the student exists and has enrolled units
                if student is None or not student.units:
                    print("Not enrolled in any Units.")
                    return  # Return to the student menu if no units are enrolled

                # Prompt for the unit name to remove
                unit_name_to_remove = input("\nEnter Unit Code to Remove: ").strip()

                # Call the method to remove the unit from the student's enrollment
                self.remove_unit_from_student(student_id, unit_name_to_remove)
            elif student_choice == '3':
                print("\nCurrent Units:"+"\n-----------------")
                self.list_student_units(student_id)
                input("\nPress any key or 'B' to return to the Menu.")
            elif student_choice == '4':
                return  # Return to main menu
            elif student_choice == '0':
                print("Exiting system.")
                sys.exit()
            else:
                print("Invalid choice.")  # Handle invalid choice

    def admin_menu(self):
        """
        Provides the menu for administrator interactions.
        """
        while True:
            self.display_header("Administrator Menu")
            print("\nAdministrator Menu:")
            print("1. Add Student")
            print("2. Remove Student")
            print("3. Add Unit")
            print("4. Remove Unit")
            print("5. List All Students")
            print("6. List All Units")
            print("7. Back to Main Menu")
            print("8. Add Unit to Student")  # New option
            print("9. Remove Unit from Student")  # New option
            print("0. Exit")
            admin_choice = input("Enter choice: ").strip()

            if admin_choice == '1':
                self.display_header("Admin Menu: Add New Student")
                new_student_name = input("Enter New Student Name: ").strip()
                new_student_id = input("Enter New Student ID: ").strip()
                new_student_address = input("Enter New Student Address: ").strip()
                new_student_phone = input("Enter New Student Phone: ").strip()
                new_student_email = input("Enter New Student Email: ").strip()
                new_student_type = input("Enter Type of Student - [F] Full-time or [P] Part-time: ").strip().upper()

                if new_student_type == 'F':
                    new_student = FullTimeStudent(new_student_id, new_student_name, new_student_address, new_student_phone, new_student_email)
                elif new_student_type == 'P':
                    new_student = PartTimeStudent(new_student_id, new_student_name, new_student_address, new_student_phone, new_student_email)
                else:
                    print("Invalid Option. Returning to Main Menu.")
                    continue

                self.add_student(new_student)
            elif admin_choice == '2':
                self.display_header("Admin Menu: Remove Student")
                print("\nCurrent Students:"+"\n-----------------")
                students = list(self.get_students())
                if not students:
                    print("No students currently enrolled.")
                else:
                    for student in self.get_students().values():
                        print(student)
                    student_id_to_remove = input("\nEnter Student ID to Remove: ").strip()
                    confirmation = input(f"Confirm removal of student '{student_id_to_remove}' (Y/N): ").strip().upper()
                    if confirmation == 'Y':
                        self.remove_student(student_id_to_remove)
                    else:
                        print("Operation cancelled.")

            elif admin_choice == '3':
                self.display_header("Admin Menu: Add New Unit")
                new_unit_name = input("Enter New Unit Code: ").strip()
                new_unit_title = input("Enter New Unit title: ").strip()
                new_unit_credit_points = input("Enter Credit Points: ").strip()
                new_unit = Unit(new_unit_name, new_unit_title, new_unit_credit_points)

                confirmation = input(f"Confirm creation of unit '{new_unit_name}' (Y/N): ").strip().upper()
                if confirmation == 'Y':
                    self.add_unit(new_unit)
                else:
                    print("Operation cancelled.")

            elif admin_choice == '4':
                self.display_header("Admin Menu: Remove Unit")
                print("\nCurrent Units:"+"\n-----------------")
                units = list(self.get_units())
                if not units:
                    print("No units currently available.")
                else:
                    for unit in self.get_units():
                        print(unit)
                unit_name_to_remove = input("\nEnter Unit Code to Remove: ").strip()
                if unit_name_to_remove in self.units:
                    confirmation = input(f"Confirm removal of unit '{unit_name_to_remove}' (Y/N): ").strip().upper()
                    if confirmation == 'Y':
                        self.units.pop(unit_name_to_remove, None)
                        print(f"Unit '{unit_name_to_remove}' removed successfully.")
                    else:
                        print("Operation cancelled.")
                else:
                    print(f"Error: Unit '{unit_name_to_remove}' does not exist. Cannot remove a non-existent unit.")
            elif admin_choice == '5':
                self.display_header("Admin Menu: List All Students")
                print("\nCurrent Students:"+"\n-----------------")
                students = list(self.get_students())
                if not students:
                    print("No students currently enrolled.")
                else:
                    for student in self.get_students().values():
                        print(student)

                input("\nPress any key or 'B' to return to the Menu.")
            elif admin_choice == '6':
                self.display_header("Admin Menu: List All Units")
                print("\nCurrent Units:"+"\n-----------------")
                units = list(self.get_units())
                if not units:
                    print("No units currently available.")
                else:
                    for unit in self.get_units():
                        print(unit)
                input("\nPress any key or 'B' to return to the Menu.")
            elif admin_choice == '7':
                return  # Return to main menu
            elif admin_choice == '8':  # Add Unit to Student
                self.display_header("Admin Menu: Add Unit to Student")
                print("\nCurrent Students:"+"\n-----------------")
                students = list(self.get_students())
                if not students:
                    print("No students currently enrolled.")
                    return
                else:
                    for student in self.get_students().values():
                        print(student)
                student_id = input("\nEnter Student ID: ").strip()
                if self.student_exists(student_id):
                    print("\nAvailable Units:"+"\n-----------------")
                    for unit in self.get_units():
                        print(unit)
                    unit_name = input("\nEnter Unit Code to Enroll: ").strip()
                    unit_to_add = self.units.get(unit_name)
                    if unit_to_add:
                        self.add_unit_to_student(student_id, unit_to_add)
                    else:
                        print(f"Unit '{unit_name}' not found.")
                else:
                    print("Student not found.")
            elif admin_choice == '9':  # Remove Unit from Student
                self.display_header("Admin Menu: Remove Unit from Student")
                print("\nCurrent Students:"+"\n-----------------")
                students = list(self.get_students())
                if not students:
                    print("No students currently enrolled.")
                    return
                else:
                    for student in self.get_students().values():
                        print(student)
                student_id = input("\nEnter Student ID: ").strip()

                if self.student_exists(student_id):
                    student = self.get_students().get(student_id)  # Retrieve the student object
                    print("\nStudent's Current Units:"+"\n-----------------")
                    self.list_student_units(student_id)
                    if student.get_units():  # Check if the student has any units
                        unit_name_to_remove = input("\nEnter Unit Code to Remove: ").strip()
                        self.remove_unit_from_student(student_id, unit_name_to_remove)
                    else:
                        print("\nThis student is not enrolled in any units.")
                        return
                else:
                    print("Student not found.")
            elif admin_choice == '0':
                print("Exiting system.")
                sys.exit()
            else:
                print("Invalid choice.")  # Handle invalid choice



def main():
    """
    Main function to initialize the system, create default entries, and start the main menu.
    """
    ses = StudentEnrollmentSystem(Administrator("Admin"))

    # Create default unit and student for the system
    ses.add_unit(Unit("FIT5136","Software Engineering", 6))
    ses.add_student(FullTimeStudent("123123", "John Lee", "123 Fake Street", "1111111111", "JohnLee@fakestreet.com"))

    ses.display_main_menu()

if __name__ == "__main__":
    main()
